### PR TITLE
[fix] Use masked mean in advantage batch normalization

### LINKED ISSUE #1305: [train][bug] `normalize_advantages_dict` computes mean over padding positions
## Summary

`normalize_advantages_dict` in `ppo_utils.py` uses an unmasked `advantages.mean()` to compute the global mean, but uses a masked computation for the variance. Padding positions (where `response_mask == 0`) contaminate the mean, producing incorrect normalized advantages.

## Location

[`skyrl/backends/skyrl_train/utils/ppo_utils.py:127-144`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L127-L144)

```python
@torch.no_grad()
def normalize_advantages_dict(data: TrainingInputBatch) -> TrainingInputBatch:
    advantages: Float[torch.Tensor, "batch_size seqlen"] = data["advantages"]
    response_masks: Float[torch.Tensor, "batch_size seqlen"] = data["response_mask"]
    num_actions: float = response_masks.sum()
    # mean
    mean: float = advantages.mean()                                    # BUG: includes padding
    # std
    std: float = ((advantages - mean).pow(2) * response_masks).sum()   # correct: masked
    rstd: float = (std / num_actions).clamp(min=1e-8).rsqrt()

    data["advantages"] = (advantages - mean) * rstd
    return data
```

Line 138 computes the mean over **all** tensor elements. Line 140 correctly masks with `response_masks`. The mean and variance use inconsistent masking.

## Call path

1. **Training loop** ([`trainer.py:271`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/train/trainer.py#L271)): `self.compute_advantages_and_returns(training_input)`
2. **Dispatch** ([`ppo_utils.py:1190-1214`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1190-L1214)): `compute_advantages_and_returns()` calls the registered advantage estimator (GAE, GRPO, RLOO, or REINFORCE++)
3. **Normalization** ([`trainer.py:277-278`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/train/trainer.py#L277-L278)): if `cfg.trainer.algorithm.advantage_batch_normalize` is enabled:
   ```python
   training_input = normalize_advantages_dict(training_input)
   ```

The same path exists in [`fully_async_trainer.py:515-516`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/train/fully_async_trainer.py#L515-L516).

## Severity by estimator

The impact depends on what the advantage estimator leaves in padding positions:

| Estimator | Padding positions | Impact |
|---|---|---|
| **GAE** ([`ppo_utils.py:1128`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1128)) | **Non-zero garbage** — [`masked_whiten`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L165-L171) normalizes using masked stats but does not zero out padding positions | **High** — mean is contaminated by arbitrary values from the GAE recurrence |
| **GRPO** ([`ppo_utils.py:1096`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1096)) | Zero — explicitly multiplies by `response_mask` | **Low-moderate** — zeros dilute the mean toward zero proportional to padding ratio |
| **RLOO** ([`ppo_utils.py:1096`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1096)) | Zero — explicitly multiplies by `response_mask` | **Low-moderate** — same dilution effect |
| **REINFORCE++** ([`ppo_utils.py:1043`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1043)) | Zero — explicitly multiplies by `response_mask` | **Low-moderate** — same dilution effect |

For the zero-padding estimators, including padding zeros still biases the mean: if the true masked mean is positive, the diluted mean is closer to zero, so the subtracted mean is too small. The bias scales with `padding_tokens / total_tokens`.

## Suggested fix

Use the already-available [`masked_mean`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L33) (imported from `torch_utils` at line 33):

```python
mean: float = masked_mean(advantages, response_masks)
```

This makes the mean computation consistent with the variance computation, averaging only over valid response positions.

### PR DESCRIPTION
**Pull request by discobot**
Fixes #1305.

The issue predates a refactor: `normalize_advantages_dict` has since moved byte-for-byte from `ppo_utils.py` into `RayPPOTrainer._normalize_advantages` on current main, bug intact, so the fix lands there. `FullyAsyncRayPPOTrainer` inherits the method, so the one change covers both the sync and fully-async paths.

The new regression test normalizes the same padded batch twice — once with zeros and once with garbage in the padding positions — and asserts identical results on valid positions, with masked mean 0 and masked std 1. It fails before the fix (padding garbage completely rewrites the normalized advantages on valid positions) and passes after, along with the rest of `tests/train/test_trainer.py` and `tests/backends/skyrl_train/utils/test_ppo_utils.py`.

---

> [!NOTE]
> **Medium Risk**
> Touches PPO policy training advantage scaling on every step when batch normalization is enabled; behavior change is intentional but can shift training dynamics for padded batches.
> 
> **Overview**
> Fixes **advantage batch z-score normalization** in `RayPPOTrainer._normalize_advantages` when `advantage_batch_normalize` is on (#1305).
> 
> The batch mean for normalization now uses **`masked_mean` over `response_mask`** instead of a plain tensor mean, so padding (zeros or stray GAE values) no longer skews statistics on real response tokens. **`num_actions`** is **`clamp(min=1)`** so an all-zero mask does not produce `0/0` NaNs in the variance scaling.
> 
> Adds regression tests: normalization is identical on valid tokens whether padding holds garbage, and outputs stay finite with an empty mask. Inherited trainers (e.g. fully async) pick up the same fix.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 462d39193b2437af2dfebe52d681f8df7f1a1d8b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/train/trainer.py [hunk 1]
```diff
@@ -1236,7 +1236,7 @@ def _normalize_advantages(
         # Step 1: Z-score normalization (if enabled)
         if self.cfg.trainer.algorithm.advantage_batch_normalize:
             num_actions = response_mask.sum()
```
[the remark this answers is not in the document]
Comment (discobot) [reply skyrl/train/trainer.py]: The `std / num_actions` division is pre-existing and unchanged by this PR; the line this PR does change is already guarded, since `masked_mean` clamps its denominator to `min=1.0` (torch_utils.py:180), same as the other ~20 call sites. An all-zero batch-wide `response_mask` means zero trainable tokens, which `masked_var` already treats as a hard error elsewhere.
